processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByQueryEngineV2.java

```diff
@@ -0,0 +1,392 @@
+                  {
+                    return new BaseSequence<>(
+                        new BaseSequence.IteratorMaker<Row, GroupByEngineIterator>()
+                        {
+                          @Override
+                          public GroupByEngineIterator make()
+                          {
+                            return new GroupByEngineIterator(
+                                query,
+                                config,
+                                cursor,
+                                bufferHolder.get(),
+                                keySerde,
+                                fudgeTimestamp
+                            );
+                          }
+
+                          @Override
+                          public void cleanup(GroupByEngineIterator iterFromMake)
+                          {
+                            iterFromMake.close();
+                          }
+                        }
+                    );
+                  }
+                }
+            ),
+            new Closeable()
+            {
+              @Override
+              public void close() throws IOException
+              {
+                CloseQuietly.close(bufferHolder);
+              }
+            }
+        )
+    );
+  }
+
+  private static class GroupByEngineIterator implements Iterator<Row>, Closeable
+  {
+    private final GroupByQuery query;
+    private final GroupByQueryConfig config;
+    private final Cursor cursor;
+    private final ByteBuffer buffer;
+    private final Grouper.KeySerde<ByteBuffer> keySerde;
+    private final DateTime timestamp;
+    private final DimensionSelector[] selectors;
+    private final ByteBuffer keyBuffer;
+    private final int[] stack;
+    private final IndexedInts[] valuess;
+
+    private int stackp = Integer.MIN_VALUE;
+    private boolean currentRowWasPartiallyAggregated = false;
+    private CloseableGrouperIterator<ByteBuffer, Row> delegate = null;
+
+    public GroupByEngineIterator(
+        final GroupByQuery query,
+        final GroupByQueryConfig config,
+        final Cursor cursor,
+        final ByteBuffer buffer,
+        final Grouper.KeySerde<ByteBuffer> keySerde,
+        final DateTime fudgeTimestamp
+    )
+    {
+      final int dimCount = query.getDimensions().size();
+
+      this.query = query;
+      this.config = config;
+      this.cursor = cursor;
+      this.buffer = buffer;
+      this.keySerde = keySerde;
+      this.keyBuffer = ByteBuffer.allocate(keySerde.keySize());
+      this.selectors = new DimensionSelector[dimCount];
+      for (int i = 0; i < dimCount; i++) {
+        this.selectors[i] = cursor.makeDimensionSelector(query.getDimensions().get(i));
+      }
+      this.stack = new int[dimCount];
+      this.valuess = new IndexedInts[dimCount];
+
+      // Time is the same for every row in the cursor
+      this.timestamp = fudgeTimestamp != null ? fudgeTimestamp : cursor.getTime();
```

Comment:
should this be Strings.nullOrEmpty(fudgeTimestamp) as the GroupByStrategyV2 passes an empty string in context always ?